### PR TITLE
Solves #1061 and #1656

### PR DESCRIPTION
**Pull request by lsaffre**
This solves #1061 and #1656, at least for me and as far as i tried.
See also http://lino-framework.org/blog/2015/0103.html
Sorry for an accidental changeset (and a second one to revert it).

### CODE REVIEW

**Comment by shimizukawa** [comment]
Thanks for contribution! However, the change makes errors. Please check the test result.

File: sphinx/ext/autodoc.py [hunk 1]
```diff
@@ -514,18 +515,22 @@ def process_doc(self, docstrings):
...
-            if not isinstance(self.analyzer.srcname, text_type):
-                filename = text_type(self.analyzer.srcname,
-                                     sys.getfilesystemencoding(), 'replace')
+            if not isinstance(self.analyzer.srcname, unicode):
+                filename = unicode(self.analyzer.srcname,
+                                   sys.getfilesystemencoding(), 'replace')
             else:
                 filename = self.analyzer.srcname
-            sourcename = u'%s:docstring of %s' % (filename, self.fullname)
+            return u'%s:<docstring of %s>' % (filename, self.fullname)
```
Comment (birkenfeld) [inline sphinx/ext/autodoc.py]: This is the cause of the test failures: the angle brackets are added compared to the old return value.

I think it's better readable with angle brackets, but:
- test should be adapted
- angle brackets should also be added in the next line for consistency

**Comment by lsaffre** [comment]
Thanks Georg for finding the reason (i had no idea where to start searching). Yes, brackets might be more readable, but for me it works without them.

**Comment by lsaffre** [comment]
Now the tests fail only on Python 3 where they say "nose.proxy.NameError: global name 'unicode' is not defined". Any suggestions how to solve this?

File: sphinx/ext/autodoc.py [hunk 2]
```diff
@@ -514,18 +515,22 @@ def process_doc(self, docstrings):
...
-        # set sourcename and add content from attribute documentation
+    def get_sourcename(self):
         if self.analyzer:
             # prevent encoding errors when the file name is non-ASCII
-            if not isinstance(self.analyzer.srcname, text_type):
-                filename = text_type(self.analyzer.srcname,
-                                     sys.getfilesystemencoding(), 'replace')
+            if not isinstance(self.analyzer.srcname, unicode):
+                filename = unicode(self.analyzer.srcname,
+                                   sys.getfilesystemencoding(), 'replace')
```
Comment (shimizukawa) [inline sphinx/ext/autodoc.py]: This 3 lines change is a cause of test failing on Py3. `text_type` is a compatibility type for py2 `unicode` and py3 `str`. please revert this 3 lines.

**Comment by lsaffre** [comment]
Thanks, Takayuki, for your hint!

**Comment by birkenfeld** [comment]
Looks good to me now.

**Comment by birkenfeld** [comment]
BTW, now with git it's easier to remove unwanted changesets. Just to a `git rebase -i`, remove the offending commits, and then push with `-f`. Github will automatically update the pull request.

**Comment by lsaffre** [comment]
Thanks, Georg, for your hint about |git rebase|. I tried it but without success... must say that this would have been my first rebase... details in my blog http://lino-framework.org/blog/2015/0105.html

**Comment by birkenfeld** [comment]
Well, looking at your blog entry, there are two snags:
- git needs EDITOR to return only when the edit is completed. I don't think this can be changed, but I assume that it's possible to set a different editor executable.
- the argument to the rebase is the head of the commits that should not be touched. Commits between that and HEAD will be offered for rebase. When you do `rebase -i master` on the master branch, it's a noop since there are no commits between master and master :) Usually the argument will be `origin/master` (rebase/edit commits made relative to upstream repository's branch), or a different branch (rebase commits onto the specified branch).

**Comment by birkenfeld** [comment]
Merged manually, thanks!

**Comment by birkenfeld** [comment]
See https://github.com/sphinx-doc/sphinx/commit/21b838454f4bb5b6a246b85cb2f521deb7552505

Note that it retained your author info, another nice bit in git.

**Comment by lsaffre** [comment]
Nice, I am part of Sphinx's history :-) Thanks, Georg and Takayuki,  for
your help and sorry for my slowness (so many other urgent projects).

Luc

**Pull request closed without merging**